include/srf/experimental/modules/module_registry.hpp

```diff
@@ -0,0 +1,123 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2021-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <nlohmann/json.hpp>
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+
+namespace srf::modules {
+
+class SegmentModule;
+
+/**
+ * Simple, thread safe, global module registry.
+ */
+class ModuleRegistry
+{
+  public:
+    using module_constructor_t =
+        std::function<std::shared_ptr<srf::modules::SegmentModule>(std::string module_name, nlohmann::json config)>;
+
+    using module_registry_map_t  = std::map<std::string, module_constructor_t>;
+    using module_namespace_map_t = std::map<std::string, module_registry_map_t>;
+    using module_name_map_t      = std::map<std::string, std::vector<std::string>>;
+
+    ModuleRegistry() = delete;
+
+    /**
+     * Return true if the registry contains 'name' else false
+     * @param name Name of the module
+     * @return boolean indicating if the registry contains the required module.
+     */
+    static bool contains(const std::string& name, const std::string& registry_namespace = "default");
+
+    /**
+     * Return true if the registry contains the namespace, else false
+     * @param registry_namespace Namespace name
+     * @return boolean indicating if the registry contains the required namespace.
+     */
+    static bool contains_namespace(const std::string& registry_namespace);
+
+    /**
+     * Attempt to retrieve the module constructor for a given module name; throws an error
+     * if the given module does not exist.
+     * @param name Name of the module
+     * @return Module constructor
+     */
+    static module_constructor_t find_module(const std::string& name, const std::string& registry_namespace = "default");
```

Comment:
I dont love this function being public. Users should be registering modules with the registry, and then creating them with `Builder::load_module_from_registry()`. When you use `find_module`, you have to use it in a specific way that could be problematic. I can easily see user doing the following (from python):
```
module = ModuleRegistry.find_module("mod_name")
builder.init_module(module)
```
Instead, users should just use `load_module_from_registry` which does all of this for you. Having access to the constructor doesnt provide any benefit since you still get the object back from `load_module_from_registry`

Reply:
I have mixed feelings on this. Generally speaking, I agree with you from a user perspective, although both approaches are equivalent functionally.

This is nice for unit testing, because you can test object registration, construction, and basic components without having a builder, reducing coupling, and without having to friend some kind of testing object.

Reply:
Based on conversation, updated the API to be `get_module_constructor`; reduces confusion as to how the object is used and improves overall readability.